Implement exception logging

aysunrhn opened the pull request:
Implemented an exception logging system for SedTRAILS that captures and logs exceptions throughout the simulation run. I added an example log file in the `examples` folder.

Closes #251 

The logger module has the features:
- Global Exception Handler: Automatically captures and logs unhandled exceptions
- Logger Manager: Centralized logging configuration with directory management
- Exception Logging Functions: log_exception() and log_simulation_state() for structured error reporting

I also added unit tests for individual logging tasks such as:
- Test core logging functionality, creating log files successfully
- Test that global exception handler works as intended, and they keyboard interruption (Ctrl+C) is ignored
- Test that logger is integrated with the Simulation class

## Relevant files
- `src/sedtrails/simulation.py`
- `tests/logger/test_logger.py`
- `examples/log_example.txt`

aysunrhn committed:
Implement exception logging

aysunrhn committed:
Add tests for the logger module

aysunrhn committed:
Initialize logger after config file is read

aysunrhn committed:
Set global exception logging in Simulation class

aysunrhn committed:
Add tests for logger module

aysunrhn committed:
Update example log file log_example.txt

aysunrhn committed:
Fix ruff errors

manuGil commented on src/sedtrails/simulation.py:
```diff
@@ -12,17 +12,31 @@
+    """Setup global exception logging for unhandled exceptions."""
+    original_excepthook = sys.excepthook
+    
+    def exception_handler(exc_type, exc_value, exc_traceback):
+        # Don't log KeyboardInterrupt (Ctrl+C)
+        if issubclass(exc_type, KeyboardInterrupt):
+            original_excepthook(exc_type, exc_value, exc_traceback)
+            return
```
Do you think we need this in for our use case? I think we want to log the interruptions, but maybe you have a different idea.

manuGil replied:
@aysunrhn was this possible? If not that's ok, we can discuss with the team if this is useful or not

manuGil commented on src/sedtrails/simulation.py:
```diff
@@ -12,17 +12,31 @@
 from sedtrails.data_manager import DataManager
 from sedtrails.particle_tracer.timer import Time, Duration, Timer
 from sedtrails.logger.logger import log_simulation_state, log_exception, _logger_manager
```
my understanding is that `_variable_name` is a convention to indicate private methods/attributes in Python. But here, we are treading `_logger_manager` as public. Could you please double check the code in logger.py is being used as intended? 
Besides, I would expect that we only need to import a single object from logger, and use that thru the script.

manuGil commented on src/sedtrails/simulation.py:
```diff
@@ -179,23 +204,28 @@ def run(self):
+
+            # Ensure logger directory is set from config
+            output_dir = self._controller.get('folder_settings.output_dir', 'results')
+            _logger_manager.log_dir = output_dir
```
This is also in line 70. Why do we need it here as well?

manuGil commented on src/sedtrails/logger/logger.py:
let's have a discussion on your implementation of this. I do not understand some of the codding conventions you used.

manuGil commented on tests/logger/test_logger.py:
```diff
@@ -0,0 +1,523 @@
+from sedtrails.exceptions.exceptions import ConfigurationError
+
+# Test Configuration Data
+def get_valid_config(output_dir):
```
Unit-test for the logger shouldn't be testing the validity of the configuration parameters. This has already been tested by the module dealing with the configuration. If you use a code assistant, please do not let it make all the work. The output should always be check for consistency and logic

manuGil reviewed (changes requested):
Hello Aysun, I can see that you have implemented a global way to log exceptions. However, I have many doubts regarding the way the logger is implemented. It looks like the code and tests aren't align with the development goals. 
Let's have a discussion about this before you make more changes.

aysunrhn committed:
Remove unncessary lines of logger directory

aysunrhn committed:
Make logger_manager public variable

aysunrhn committed:
Update logger.py to initialize public logger_manager

aysunrhn committed:
Update logger.py

aysunrhn committed:
Update logger.py

aysunrhn committed:
Reduce tests to keep only relevant cases

aysunrhn committed:
Remove unncessary imports

aysunrhn committed:
Update logger.py to require logging dir for initialization

aysunrhn committed:
Update simulation.py to initialize logger

aysunrhn committed:
Update setup_global_exception_logging

aysunrhn committed:
Update test_logger.py according to changes in logger.py

aysunrhn committed:
Update LoggerManager class for more understandable structure

aysunrhn committed:
Update similation.py to be consistent with the new LoggerManager

aysunrhn committed:
Update test_logger.py to be consistent with the new LoggerManager class

aysunrhn commented:
UPDATE: following our last discussion about the logging module

- .Added `log_simulation_state()` and ` log_exception()` methods to the `LoggerManager` class in the `logger.py` file. Moved them from standalone functions into the LoggerManager class itself. This simplifies how we use the LoggerManager
```python
# Create manager and set up logger
logger_manager = LoggerManager(output_dir)
logger_manager.setup_logger()

# Use manager methods directly
logger_manager.log_simulation_state(state)
logger_manager.log_exception(e, "context")
```
- Updated `simulation.py` and `test_logger.py` to be consistent with the changes in the `logger.py`

manuGil committed:
Merge branch 'dev' into logger-exception

manuGil commented on tests/logger/test_logger.py:
```diff
@@ -0,0 +1,222 @@
+        with pytest.raises(FileNotFoundError):
+            logger_manager.setup_logger()
+
+if __name__ == '__main__':
```
Why do we need to call the script itself from main. Pytest automatically discover test files, classes and functions. Unless, this is an special case. I suggest to remove this

manuGil reviewed (commented):
@aysunrhn  Could you please read my new comments and questions, after that you can merge this. In the upcoming tasks that may involve `simulation.py`, it would be best not to modify that file. I am still working on it and when we both make changes, we have to solve conflicts. I suggest, we test other components in separated files, and we leave the integration (use) to a bit later (once I have completed the work on `simulation.py`).

aysunrhn committed:
Update test_logger.py

aysunrhn committed:
Fix ruff errors
